dbms/src/Storages/StorageKafka.cpp

```diff
@@ -242,46 +230,51 @@ BlockInputStreams StorageKafka::read(
     check(column_names);
     processed_stage = QueryProcessingStage::FetchColumns;
 
-    if (!conf)
+    if (num_consumers == 0)
         return BlockInputStreams();
 
     BlockInputStreams streams;
-    streams.reserve(num_streams);
+    streams.reserve(std::min(num_streams, num_consumers));
 
-    // Note: The block size is set to 1, otherwise it'd have to be able to return excess buffered messages
-    for (size_t i = 0; i < num_streams; ++i)
-        streams.push_back(std::make_shared<KafkaBlockInputStream>(*this, context, schema_name, 1));
+    // Claim as many consumers as requested, but don't block
+    for (size_t i = 0; i < streams.capacity(); ++i)
```

Comment:
Strange code.
It's not guaranteed, that `capacity` is equal to the value that was requested in `reserve`.

Reply:
Right, let me fix that.